### ISSUE TITLE
Not Saving Last Checkpoint After Update to 1.0.11

### ISSUE THREAD

**Issue by kmacmcfarlane** [d8ahazard/sd_dreambooth_extension#1000]
### Kindly read the entire form below and fill it out with the requested information.

**Please find the following lines in the console and paste them below. _If you do not provide this information, your
issue will be automatically closed._**

`
Python revision: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
Dreambooth revision: d4e2f7d5dc5c67cc6292c8b8137b688a43247c35
SD-WebUI revision: 0cc0ee1bcb4c24a8c9715f66cede06601bfc00c8

Checking Dreambooth requirements...
[+] torch version 1.13.1+cu117 installed.
[+] torchvision version 0.14.1+cu117 installed.
[+] bitsandbytes version 0.35.4 installed.
[+] diffusers version 0.13.1 installed.
[+] transformers version 4.26.1 installed.
[+] xformers version 0.0.17.dev464 installed.
`

**Have you read the [Readme](https://github.com/d8ahazard/sd_dreambooth_extension#readme)?**
Yes

**Have you completely restarted the stable-diffusion-webUI, not just reloaded the UI?**
Yes

**Have you updated Dreambooth to the latest revision?**
Yes, with 1.0.11 dreambooth extension release

**Have you updated the Stable-Diffusion-WebUI to the latest version?**
Yes

**No, really. Please save us both some trouble and update the SD-WebUI and Extension and restart before posting this.
Reply 'OK' Below to acknowledge that you did this.**
OK

**Describe the bug**
When doing a training of a model specifying 150 epochs with checkpoints over 25 epochs, the last checkpoint is not being saved.

The checkbox under `Saving` tab `Generate a .ckpt file when training completes.` is checked. It seems like it's compiling it, but nothing gets output to the models/Stable-Diffusion/ dir. I am also using the `Save Checkpoint to Subdirectory` and `Save separate diffusers snapshots when training completes.` options checked in that tab.

My workflow has been to create a new dreambooth model overwriting the previous to try different learning rates, blanking out the previous dir in `models/dreambooth` by pressing the `Create model` button in the gradio UI, so it's possible that's part of the bug.

As an experiment, I trained on a new dreambooth model using the same settings except with 25 epochs and outputing every 5 epochs. Surprisingly, this test output all 5 expected checkpoints. There seems to be a difference when the training is longer and the system has been under load for a while.

Occasionally, it will kill the stable-diffusion-webui with no stack track in the last steps at the end of the training run. Only `Killed` is printed in the console. Restarting stable-diffusion-webui after creating the model (and remembering to save the params, hopefully) and before starting training seems to prevent it from getting killed, I think.

I'm using the base model `rpg_V4.safetensors` from civitai with an unfrozen model type.

**Provide logs**

No errors in console output

**Environment**

What OS?
Ubuntu

If Windows - WSL or native?
WSL

What GPU are you using?
3090ti

**Screenshots/Config**
```
{
    "adamw_weight_decay": 0.01,
    "attention": "xformers",
    "cache_latents": true,
    "clip_skip": 1,
    "concepts_list": [
        {
            "class_data_dir": "/mnt/y/ai/training/bidi/classification-RPGv4",
            "class_guidance_scale": 7.5,
            "class_infer_steps": 40,
            "class_negative_prompt": "",
            "class_prompt": "[filewords]",
            "class_token": "person",
            "instance_data_dir": "/mnt/y/ai/training/bidi/preprocessed-512",
            "instance_prompt": "[filewords]",
            "instance_token": "bidi-ddim",
            "is_valid": true,
            "n_save_sample": 4,
            "num_class_images_per": 50,
            "sample_seed": -1,
            "save_guidance_scale": 7,
            "save_infer_steps": 40,
            "save_sample_negative_prompt": "",
            "save_sample_prompt": "a (close-up:1.1) photo of bidi-ddim (kneeling:1.4) in a forest with a creek running behind her, (looking at camera:1.2) ",
            "save_sample_template": ""
        }
    ],
    "concepts_path": "",
    "custom_model_name": "",
    "deis_train_scheduler": false,
    "deterministic": false,
    "ema_predict": false,
    "epoch": 150,
    "epoch_pause_frequency": 0,
    "epoch_pause_time": 0,
    "freeze_clip_normalization": false,
    "gradient_accumulation_steps": 1,
    "gradient_checkpointing": false,
    "gradient_set_to_none": false,
    "graph_smoothing": 50,
    "half_model": false,
    "train_unfrozen": true,
    "has_ema": false,
    "hflip": false,
    "infer_ema": false,
    "initial_revision": 0,
    "learning_rate": 2.7e-07,
    "learning_rate_min": 1e-06,
    "lifetime_revision": 0,
    "lora_learning_rate": 0.0002,
    "lora_model_name": "bidi-ddim-RPGv4-lion_1800.pt",
    "lora_unet_rank": 4,
    "lora_txt_rank": 4,
    "lora_txt_learning_rate": 0.0002,
    "lora_txt_weight": 1,
    "lora_weight": 1,
    "lr_cycles": 1,
    "lr_factor": 0.5,
    "lr_power": 1,
    "lr_scale_pos": 0.5,
    "lr_scheduler": "constant_with_warmup",
    "lr_warmup_steps": 0,
    "max_token_length": 75,
    "mixed_precision": "fp16",
    "model_name": "bidi-ddim-RPGv4-lion",
    "model_dir": "models/dreambooth/bidi-ddim-RPGv4-lion",
    "model_path": "models/dreambooth/bidi-ddim-RPGv4-lion",
    "num_train_epochs": 150,
    "offset_noise": 0,
    "optimizer": "Lion",
    "pad_tokens": true,
    "pretrained_model_name_or_path": "models/dreambooth/bidi-ddim-RPGv4-lion/working",
    "pretrained_vae_name_or_path": null,
    "prior_loss_scale": false,
    "prior_loss_target": 100.0,
    "prior_loss_weight": 0.75,
    "prior_loss_weight_min": 0.1,
    "resolution": 512,
    "revision": 1800,
    "sample_batch_size": 1,
    "sanity_prompt": "a photo of a tree frog",
    "sanity_seed": 420420.0,
    "save_ckpt_after": true,
    "save_ckpt_cancel": false,
    "save_ckpt_during": true,
    "save_ema": true,
    "save_embedding_every": 25,
    "save_lora_after": true,
    "save_lora_cancel": false,
    "save_lora_during": false,
    "save_preview_every": 25,
    "save_safetensors": true,
    "save_state_after": true,
    "save_state_cancel": false,
    "save_state_during": false,
    "scheduler": "DDIM",
    "shuffle_tags": true,
    "snapshot": "",
    "split_loss": true,
    "src": "/home/rt/ai/AUTOMATIC1111/stable-diffusion-webui/models/Stable-diffusion/rpg_V4.safetensors",
    "stop_text_encoder": 1,
    "strict_tokens": false,
    "tf32_enable": false,
    "train_batch_size": 1,
    "train_imagic": false,
    "train_unet": true,
    "use_concepts": false,
    "use_ema": true,
    "use_lora": null,
    "use_lora_extended": false,
    "use_subdir": true,
    "v2": false
}
```

**Comment by Arilziem** [comment]
That "killed" message would suggest it is running out of RAM I think? Check if your VM / WSL instance has enough RAM and maybe increase it a touch.

Not sure if there might also be something not getting purged from RAM after creating the model, as suggested by your experience of it failing more often when not restarting before training.

**Comment by kmacmcfarlane** [comment]
The thing is it only sometimes gets the `Killed` message, in that case it dies. The main issue I'm trying to address is it quietly not saving the final checkpoint without exiting when running a 20-30 min training. I'm using 6 training images and 50 classification images per training images. This is a 512 model.

I'll check out the WSL ram, but I have 64GB on this system and it usually hovers around 42GB used while training. There might be a hard-limit on WSL exclusive of other things on the main OS though.

I'll try some low memory dreambooth settings just to rule that out (EMA, zero gradients, etc) and see if the problem persists. It's suspicious that this began happening after updating to the new release though.

**Comment by Arilziem** [comment]
I believe WSL2 will default to half of system RAM max which should be enough, but something to check out for sure. "Killed" is something that comes from the OS so a resource issue seems likely. Trying a low memory run is a good idea.

It not saving the checkpoint is definitely weird, and if it does not complain about anything then there is probably something else going on (i.e. not RAM related).

When you say it does not create a checkpoint - do you mean that it does not create a file? Or do you mean it does not overwrite an existing checkpoint?

**Comment by kmacmcfarlane** [comment]
It doesn't write a file whether or not there are already files with the same name in the subdir. It seems to be overwriting existing files with the same name as expected.

**Comment by kmacmcfarlane** [comment]
I did some more tests after restarting my PC, and I think I can attribute both issues to creeping resource issues like you suggested based on that. Updating to the new version may have just been a coincidence? Memory reported by the VM os does get close to half system RAM, so I think that's the underlying issue.

**Comment by Arilziem** [comment]
Ok great. Resource use seems a bit high-ish, but tbh I don't know what's normal since I'm always doing native training with 64GB, so this has never been an issue for me.

What still feels off to me is the 'silently not generating a checkpoint' thing.

I did notice that `"use_lora"` is `null` instead of `false` in the training configuration you posted (which should not happen, it should either be `true` or `false`). There has been another issue (#991) where for some reason the configuration contained a wrong value. Since python does not have `null` (but rather `NoneType`) there might be something weird happening with `null` evaluating `True` maybe? This is pure speculation though.

If you feel like investigating a bit, you could try checking and unchecking use_lora (or manually edit `db_config.json` and load settings) if this happens again. Or, if everything is going smooth now, whether the value has changed to `false`.

Other than that, if there are no more issues - have fun experimenting. :)

**Comment by kmacmcfarlane** [comment]
Everything's been working correctly after the reboot. I suspect simply restarting WSL would do the trick without rebooting the whole system. Currently the `use_lora` key is still set to `null` so that doesn't seem to be the issue. Thanks for talking through this with me :-)
